src/operator/adapter.rs

```diff
@@ -0,0 +1,177 @@
+/// [module-level documentation](`crate::operator::adapter`) for details.
+pub struct UnaryOperatorAdapter<I, O, Op> {
+    op: Op,
+    _types: PhantomData<(I, O)>,
+}
+
+impl<I, O, Op> Operator for UnaryOperatorAdapter<I, O, Op>
+where
+    Op: Operator,
+    I: 'static,
+    O: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        self.op.name()
+    }
+    fn clock_start(&mut self, scope: Scope) {
+        self.op.clock_start(scope);
+    }
+    fn clock_end(&mut self, scope: Scope) {
+        self.op.clock_end(scope);
+    }
+    fn is_async(&self) -> bool {
+        self.op.is_async()
+    }
+    fn ready(&self) -> bool {
+        self.op.ready()
+    }
+    fn register_ready_callback<F>(&mut self, cb: F)
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        self.op.register_ready_callback(cb);
+    }
+}
+
+impl<RI, I, RO, O, Op> UnaryOperator<RI, RO> for UnaryOperatorAdapter<I, O, Op>
+where
+    Op: UnaryOperator<I, O>,
+    RI: SharedRef<I>,
+    I: 'static,
+    RO: From<O>,
+    O: 'static,
+{
+    fn eval(&mut self, input: &RI) -> RO {
+        RO::from(self.op.eval(input.borrow()))
+    }
+
+    fn eval_owned(&mut self, input: RI) -> RO {
+        RO::from(match input.try_into_owned() {
+            Ok(v) => self.op.eval_owned(v),
+            Err(v) => self.op.eval(v.borrow()),
+        })
+    }
+
+    fn input_preference(&self) -> OwnershipPreference {
+        self.op.input_preference()
+    }
+}
+
+/// Binary operator adapter unwraps input values of types
+/// `I1` and `I2` wrapped in shared references.  See
+/// [module-level documentation](`crate::operator::adapter`) for details.
+pub struct BinaryOperatorAdapter<I1, I2, O, Op> {
+    op: Op,
+    _types: PhantomData<(I1, I2, O)>,
+}
+
+impl<I1, I2, O, Op> BinaryOperatorAdapter<I1, I2, O, Op> {
+    pub fn new(op: Op) -> Self {
+        Self {
+            op,
+            _types: PhantomData,
+        }
+    }
+}
+
+impl<I1, I2, O, Op> Operator for BinaryOperatorAdapter<I1, I2, O, Op>
+where
+    Op: Operator,
+    I1: 'static,
+    I2: 'static,
+    O: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        self.op.name()
+    }
+    fn clock_start(&mut self, scope: Scope) {
+        self.op.clock_start(scope);
+    }
+    fn clock_end(&mut self, scope: Scope) {
+        self.op.clock_end(scope);
+    }
+    fn is_async(&self) -> bool {
+        self.op.is_async()
+    }
+    fn ready(&self) -> bool {
+        self.op.ready()
+    }
+    fn register_ready_callback<F>(&mut self, cb: F)
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        self.op.register_ready_callback(cb);
+    }
+}
+
+impl<RI1, I1, RI2, I2, RO, O, Op> BinaryOperator<RI1, RI2, RO>
+    for BinaryOperatorAdapter<I1, I2, O, Op>
+where
+    Op: BinaryOperator<I1, I2, O>,
+    RI1: SharedRef<I1>,
+    RI2: SharedRef<I2>,
+    I1: 'static,
+    I2: 'static,
+    RO: From<O>,
+    O: 'static,
+{
+    fn eval(&mut self, lhs: &RI1, rhs: &RI2) -> RO {
```

Comment:
usually lhs is left-hand side, I think left and right are more appropriate, but it's not a big deal.